### DISCUSSION TITLE
Completion not working for certain APIs, e.g. Canvas, XMLHttpRequest, etc.

### DISCUSSION THREAD

**Discussion by nathanvy** [typescript-language-server/typescript-language-server#730] [Q&A]
Hi,

Let me preface this with a caveat that I'm not super fluent in all the terminology around language servers (i.e. I don't know what a "code lens" is), so please be gentle if I'm using the wrong terms for things.

I'm writing an application using vanilla js + WebGL and my development stack is:

- emacs
- lsp-mode
- company-mode
- ts-ls installed through emacs

Completion works fine for the core parts of javascript, e.g. `getElementByID()` and friends.  But if I want to call into, let's say, the [canvas](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext) API, I don't have completion working.  Here's an example:

```javascript
var canvas = document.getElementById("foo");
var ctx = canvas.getContext("bar");
```
<img width="408" alt="Screenshot 2023-06-14 at 1 55 20 PM" src="https://github.com/typescript-language-server/typescript-language-server/assets/2128134/50803c2d-6be0-40f6-97df-0362eecfc19d">

Note how the `getContext` isn't offered by the language server as a completion option.

Here's a second example:

```javascript
xmlr = new XMLHttpRequest();
xmlr.open('GET', url, false);
```

Similarly, the `open` isn't offered as a completion option.

How do I fix this?

**Comment by rchl** [comment]
This is controlled through `tsconfig.json`. See https://www.typescriptlang.org/tsconfig#lib

**Comment by nathanvy** [comment]
Thanks for the reply.  I've included a `tsconfig.json` file at the root of my project containing the following:

```json
{
  "compilerOptions": {
      "target": "es2020",
      "lib": ["es2020"],
      "module": "commonjs"
   }
}

```

It did not appear to have any effect whatsoever.  I still don't get completion on e.g. `canvas.getCo` hoping for `.getContext()`.  Have I missed a step?

**Reply by guillaumebrunerie** [reply]
You need to add "dom" as a library as well. The Canvas API is not part of ECMAScript 2020, it's part of the DOM specification.
